File: 2026-02/src/day_05/boj_13913/Main.java
package day_05.boj_13913;

import java.io.*;
import java.util.*;

public class Main {

    static final int MAX = 100000;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());

        System.out.println(bfs(n, k));
    }

    static String bfs(int n, int k) {
        Queue<Integer> q = new ArrayDeque<>();
        q.offer(n);

        boolean[] vis = new boolean[1 + MAX];
        vis[n] = true;

        int dist = 0;

        int[] prv = new int[1 + MAX];
        Arrays.fill(prv, -1);

        while (!q.isEmpty()) {
            int sz = q.size();

            while (sz-- > 0) {
                int cur = q.poll();
                if (cur == k) return traceback(dist, k, prv);

                int nxt1 = cur - 1;
                if (nxt1 >= 0 && !vis[nxt1]) {
                    q.offer(nxt1);
                    vis[nxt1] = true;
                    prv[nxt1] = cur;
                }

                int nxt2 = cur + 1;
                if (nxt2 <= MAX && !vis[nxt2]) {
                    q.offer(nxt2);
                    vis[nxt2] = true;
                    prv[nxt2] = cur;
                }

                int nxt3 = cur * 2;
                if (nxt3 <= MAX && !vis[nxt3]) {
                    q.offer(nxt3);
                    vis[nxt3] = true;
                    prv[nxt3] = cur;
                }
            }

            dist++;
        }

        return null;
    }

    static String traceback(int dist, int k, int[] prv) {
        StringBuilder sb = new StringBuilder();
        Deque<Integer> stack = new ArrayDeque<>();

        sb.append(dist).append("\n");

        int pos = k;
        while (pos != -1) {
            stack.push(pos);
            pos = prv[pos];
        }

        while (!stack.isEmpty()) {
            sb.append(stack.pop()).append(" ");
        }

        return sb.toString();
    }
}

File: 2026-02/src/day_05/boj_13913/Main.cpp
#include <bits/stdc++.h>
using namespace std;

const int MAX = 100000;
bool vis[1 + MAX];
int prv[1 + MAX];

void traceback(int dist, int k) {
    vector<int> v;

    int pos = k;
    while (pos != -1) {
        v.push_back(pos);
        pos = prv[pos];
    }

    reverse(v.begin(), v.end());

    cout << dist << '\n';

    for (int x : v) {
        cout << x << ' ';
    }
}

void bfs(int n, int k) {
    queue<int> q;
    q.push(n);

    vis[n] = true;

    int dist = 0;

    memset(prv, -1, sizeof(prv));

    while (!q.empty()) {
        int sz = q.size();

        while (sz-- > 0) {
            int cur = q.front();
            q.pop();

            if (cur == k) {
                traceback(dist, k);
                return;
            }

            for (int nxt : {cur - 1, cur + 1, cur * 2}) {
                if (0 <= nxt && nxt <= MAX && !vis[nxt]) {
                    q.push(nxt);
                    vis[nxt] = true;
                    prv[nxt] = cur;
                }
            }
        }

        dist++;
    }
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    int n, k;
    cin >> n >> k;

    bfs(n, k);
}
